### PR TITLE
Port polyfilter

### PR DESCRIPTION
**Pull request by keskitalo**
Port the 1 and 2D polyfilters. Upgrade the 2D filter to run on subsets of the focalplane.

### CODE REVIEW

File: src/toast/ops/polyfilter.py [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+            t_clean = 0
+
+            # communicator for processes with the same sample range
+            comm = obs.comm_col
+
+            # Detectors to process
+
+            detectors = []
+            for det in obs.all_detectors:
+                if pat.match(det) is None:
```
Comment (tskisner) [inline src/toast/ops/polyfilter.py:115]: The `Pipeline` operator can be used to execute multiple sub-operators one "detector set" at a time (not just one detector at a time).  However, we have explicitly banned that mode of operation above (we don't allow calling `exec()` with a subset of detectors).  So if we want to filter det sets (as defined in each observation), we would need to use some other method of selecting each detector set (either a focalplane key or a regex that matches their names).  I guess this is fine for now, but at some point we might want to add extra logic to check that if a subset of detectors is given, *AND* if it is a detector set, then allow it.

File: src/toast/ops/polyfilter.py [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+
+    def _finalize(self, data, **kwargs):
+        return
+
+    def _requires(self):
+        req = {
+            "meta": list(),
+            "shared": [self.shared_flags],
+            "detdata": [self.det_data, self.det_flags],
+            "intervals": [self.view],
```
Comment (tskisner) [inline src/toast/ops/polyfilter.py:329]: We should only append the flag and interval names if they are not `None`.

File: src/toast/ops/polyfilter.py [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+        }
```
Comment (tskisner) [inline src/toast/ops/polyfilter.py:456]: Same note here about only appending flags / view if they are not None.

**Review by tskisner — APPROVED** [review]
Just some minor comments, but thanks for porting all this!

File: src/toast/ops/polyfilter.py [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+    @function_timer
+    def _exec(self, data, detectors=None, **kwargs):
+        """Apply the 2D polynomial filter to the signal.
+
+        Args:
+            data (toast.Data): The distributed data.
+
+        """
+        if detectors is not None:
+            raise RuntimeError("PolyFilter2D cannot be run in batch mode")
```
[the remark this answers is not in the document]
Comment (keskitalo) [reply src/toast/ops/polyfilter.py:95]: Sure.

File: src/toast/ops/polyfilter.py (hunk 2 above)
Comment (keskitalo) [reply src/toast/ops/polyfilter.py:329]: Done.

File: src/toast/ops/polyfilter.py (hunk 3 above)
Comment (keskitalo) [reply src/toast/ops/polyfilter.py:456]: Done.

**Comment by keskitalo** [comment]
Addressed the feedback. Tests pass. Merging.

**Pull request merged**